io/io/src/TFileMergeServer.cxx

```diff
@@ -0,0 +1,348 @@
+
+      for (unsigned int f = 0; f < fClients.size(); ++f) {
+         if (fClients[f].fFile) {
+            R__DeleteObject(fClients[f].fFile, true);
+         } else {
+            // We back up the file (probably due to memory constraint)
+            TFile *file = TFile::Open(fClients[f].fLocalName, "UPDATE");
+            // Remove object that can be incrementally merged and
+            // will be reset by the client code.
+            R__DeleteObject(file, true);
+            file->Write();
+            delete file;
+         }
+      }
+      fLastMerge = TTimeStamp();
+      fNClientsContact = 0;
+      fClientsContact.Clear();
+
+      return result;
+   }
+
+   void RegisterClient(UInt_t clientId, TFile *file)
+   {
+      ++fNClientsContact;
+
+      fClientsContact.SetBitNumber(clientId);
+
+      if (fClients.size() < clientId + 1) {
+         fClients.push_back(ClientInfo(fFilename, clientId));
+      }
+      fClients[clientId].Set(file);
+   }
+};
+
+} // unnamed namespace
+
+namespace ROOT {
+namespace Experimental {
+
+std::shared_ptr<TFileMergeClient> TFileMergeServer::GetClient()
+{
+   std::shared_ptr<TFileMergeClient> client(new TFileMergeClient(*this));
+   fClients.push_back(client);
+   return client;
+}
+
+void TFileMergeServer::Listen()
+{
+   std::unique_lock<std::mutex> wlock(fWriteMutex);
+
+   THashTable mergers;
+   bool cache = false;
+
+   static Int_t clientId = 0;
+
+   while (!finish) {
+      fCV.wait(wlock, [this](){ return !this->fQueue.empty(); });
+
+      while (!fQueue.empty()) {
+         TBufferFile *buffer;
+
+         {
+            std::lock_guard<std::mutex> qlock(fQueueMutex);
+            buffer = fQueue.front();
+            fQueue.pop();
+         }
+
+         buffer->SetReadMode();
+         buffer->SetBufferOffset();
+
+         Long64_t length;
+         TString filename;
+
+         buffer->ReadTString(filename);
+         buffer->ReadLong64(length);
+
+         // UPDATE because we need to remove the TTree after merging them.
+         TMemFile *transient = new TMemFile(filename, buffer->Buffer() + buffer->Length(), length, "UPDATE");
+
+         buffer->SetBufferOffset(buffer->Length() + length);
+
+         // control how often the histogram are merged.  Here as soon as half the clients have reported.
+         const Float_t clientThreshold = 0.75;
+
+         ThreadFileMerger *info = (ThreadFileMerger *)mergers.FindObject(filename);
+
+         if (!info) {
+            info = new ThreadFileMerger(filename, cache);
+            mergers.Add(info);
+         }
+
+         if (R__NeedInitialMerge(transient)) {
+            info->InitialMerge(transient);
+         }
+
+         info->RegisterClient(clientId, transient);
+
+         if (info->NeedMerge(clientThreshold))
+            info->Merge();
+
+         transient = nullptr;
+         delete buffer;
+      }
+   }
+
+   TIter next(&mergers);
+   ThreadFileMerger *info;
+   while ((info = (ThreadFileMerger *)next())) {
```

Comment:
ThreadFileMerger will be removed altogether later, in favor of using TFileMerger directly in TFileMergeServer.